### PR TITLE
Refactors and optimizes breath code (Saves 12% of carbon/Life())

### PR DESCRIPTION
**Pull request by LemonInTheDark**
### How things work

As things currently stand, when a mob breaths several things happen (simplified to focus on the stupid)

We assert the existance of all possible breathable gases, and pull partial pressures for them
Then we walk through all possible interactions lungs could have with these gases, one by one, and see if they're happening or not
As we go we are forced to cleanup potential alerts caused by the previous breath, even if those effects never actually happen
At the end we clear out all the unused gas ids, and handle the temperature of the breath.

### What sucks

There's I'd say 3 different types of gas reactions.

- You can "need" a gas to survive. o2, n2 and plasma all fall into this category
- A gas can do something to you while it's in your system. This applies to most gas types
- Variation on the previous, some gases do cleanup when they're not in your system, or when there isn't much of them in the first place

The main headache here is that second one, constantly cleaning up potential side effects sucks, and fixing it would require a lot of dummy variables

There's other suckage too.

Needing to constantly check for a gas type even if it isn't there is stupid, and leads to wasted time It's also really annoying to do subtypes in this system. 
There is what amounts to a hook proc you can override, but you can't override the reaction to a gas type.
It also just like, sucks to add new gases. one mega proc smells real stupid.

### Improvements

In the interest of speed:

- I'd like to build a system that doesn't require manually checking for gas
- Reacting to gas "disappearing" should be promoted by the system, instead of being hacky.
- I would like to avoid needing to assert the existence of all possible gases, as this is slow on both the assert and the garbage collect.

In the interest of dev ergonomics:

- It should be easy to define a new gas reaction 
- It should be easy for subtypes to implement their own gas reactions. The current method of vars on the lung is all tangled up and not really undoable as of now, but I'd like to not require it
- It should be possible to fully override how a gas is handled

### What I've Done

Lungs have 3 lists of proc paths stored on them

Each list handles a different way the lung might want to interact with a gas.
There's a list for always processing on a gas (we use this for stuff that's breathed), a list for handling a gas in our breath, and a list for reacting to a gas previously being in our breath, but not any more.

Lungs fill out these lists using a helper proc during Initialize()
Then, when it comes time to breath, we loop over the gas in the breath and react to it.
We also keep track of the previous list of partial pressures, which we calculate for free here, and use that to figure out when to call the loss reactions.

This proc pattern allows for overrides, easy reactions to removals, lower indentation code and early returns, and better organization of signal handlers

It's also significantly faster. Ballpark 4x faster

### Misc

Removes support for breathing co2, and dying from n2 poisoning. 
They were both unused, and I think it's cringe to clutter these procs even further

Added "do we even have oxyloss" checks to most cases of passive breathing. 
This is a significant save, since redundant adjustoxy's are decently expensive at the volume of calls we have here.

Fixes a bug with breathing out if no gas is passed in, assigning a var to another var doesn't perform a copy

Rewrote breathe_gas_volume() slightly to insert gas into an immutable mix stored on the lung, rather then one passed in
This avoids passing of a gas_mixture around just to fill a hole. 

I may change my mind on this, since it would be nice to have support for temperature changing from a hot/cold breath. 
Not gonna be done off bodytemp tho lord no.

Uses merge() instead of a hard coded version to move the gas ids over. 
This is slightly slower with lower gas counts but supports more things in future and is also just easier to read.

## Why It's Good For The Game

Faster, easier to work with and read (imo)

Profiles: 
[breath_results_old.txt](https://github.com/tgstation/tgstation/files/11068247/breath_results_old.txt)
[breath_results_pre_master.txt](https://github.com/tgstation/tgstation/files/11068248/breath_results_new.txt)
[breath_results_new.txt](https://github.com/tgstation/tgstation/files/11068349/breath_results_new.txt)

(These profiles were initially missing #73026. Merging this brings the savings from 16% to 12%. Life is pain)

### CODE REVIEW

**Comment by Floofies** [comment]
Great job making lungs modular. I like how you deflated `check_breath` so we can at least start to organize properly. I don't mind what this PR breaks because the changes in it are a big step forward for organizing the massive amount of loosely-written code. The proc conga line is beautiful and the way you are using `add_gas_reaction` really helps to organize things.

Also based on what I read earlier it looks like we need to adjust the unit test to have some "baked-in" expectations instead of reading from the lung variables for min/max gas directly. Realistically the fact that this PR broke the test so easily is a good indicator

File: code/modules/surgery/organs/lungs.dm [hunk 1]
```diff
@@ -92,16 +105,50 @@
...
 	if(safe_oxygen_min)
 		respiration_type |= RESPIRATION_OXYGEN
 	if(safe_plasma_min)
 		respiration_type |= RESPIRATION_PLASMA
 
+	// Sets up what gases we want to react to, and in what way
+	// always is always processed, while_present is called when the gas is in the breath, and on_loss is called right after a gas is lost
+	// The naming convention goes like this
+	// always : breath_{gas_type}
+	// safe/nuteral while_present : consume_{gas_type}
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:121]:
```suggestion
	// safe/neutral while_present : consume_{gas_type}
```

File: code/modules/surgery/organs/lungs.dm [hunk 2]
```diff
@@ -92,16 +105,50 @@
...
+	// safe/nuteral on_loss : lose_{gas_type}
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:122]:
```suggestion
	// safe/neutral on_loss : lose_{gas_type}
```

File: code/modules/surgery/organs/lungs.dm [hunk 3]
```diff
@@ -113,10 +160,389 @@
 	receiver.clear_alert(ALERT_NOT_ENOUGH_PLASMA)
 	receiver.clear_alert(ALERT_NOT_ENOUGH_N2O)
 
+/obj/item/organ/internal/lungs/Remove(mob/living/carbon/organ_owner, special)
+	. = ..()
+	// This is very "manuel" I realize, but it's useful to ensure cleanup for gases we're removing happens
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:165]:
```suggestion
	// This is very "manual" I realize, but it's useful to ensure cleanup for gases we're removing happens
```

unless you're being funny

File: code/modules/surgery/organs/lungs.dm [hunk 4]
```diff
@@ -113,10 +160,389 @@
...
+ * proc/always * Always called. Best suited for breathing procs, like oxygen
+ *
+ * while_present and always get the same arguments (mob/living/carbon/breather, datum/gas_mixture/breath, pp, old_pp)
+ * on_loss is almost exactly the same, except it doesn't pass in a current partial pressure, since one isn't avalible
+ */
+/obj/item/organ/internal/lungs/proc/add_gas_reaction(gas_type, while_present = null, on_loss = null, always = null)
+	if(always)
+		breathe_always[gas_type] = always
+
+	if(!while_present && !on_loss)
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:192]:
```suggestion
	if(isnull(while_present) && isnull(on_loss))
```

just based on the arg-defaults for this proc

File: code/modules/surgery/organs/lungs.dm [hunk 5]
```diff
@@ -113,10 +160,389 @@
...
+		return
+
+	if(while_present)
+		breath_present[gas_type] = while_present
+	if(on_loss)
+		breath_lost[gas_type] = on_loss
+
+/// Return this from a while_present proc to call its on_loss version, if one exists
+/// Useful for doing "we're done" effects without duped code
+#define BREATH_LOST 1
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:202]: did you undef this? not sure if github is shitting itself viewing this file but i don't see it

File: code/modules/surgery/organs/lungs.dm [hunk 6]
```diff
@@ -113,10 +160,389 @@
...
+		else
+			n2o_euphoria = EUPHORIA_INACTIVE
+		return
+
+	// More N2O, more severe side-effects. Causes stun/sleep.
+	if(old_n2o_pp < n2o_para_min)
+		breather.throw_alert(ALERT_TOO_MUCH_N2O, /atom/movable/screen/alert/too_much_n2o)
+	n2o_euphoria = EUPHORIA_ACTIVE
+
+	// 60 gives them one second to wake up and run away a bit!
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:493]: no need to have this comment anymore because it's no longer correct anyways (did they really not know how deciseconds worked???)

File: code/modules/surgery/organs/lungs.dm [hunk 7]
```diff
@@ -128,512 +554,125 @@
...
-	// Activate mood on first flag, remove on second, do nothing on third.
 
 	if(has_moles)
 		handle_breath_temperature(breath, breather)
-		// Transfer exchanged gases into breath for exhalation.
-		for(var/gas_type in breath_gases_out)
-			breath.assert_gas(gas_type)
-			breath_gases[gas_type][MOLES] += breath_gases_out[gas_type][MOLES]
+		// Merge breath_out into breath. They're kept seprerate before now to ensure stupid like, order of operations shit doesn't happen
+		// But that time has past
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:651]:
```suggestion
		// But that time has passed
```

or

```suggestion
		// But that time is now in the past
```

File: code/modules/surgery/organs/lungs.dm [hunk 8]
```diff
@@ -655,7 +694,7 @@
 	if(breath_pp)
 		var/ratio = safe_breath_min / breath_pp
 		suffocator.adjustOxyLoss(min(5 * ratio, HUMAN_MAX_OXYLOSS))
-		return true_pp * ratio / 6
+		return mole_count * ratio / 6
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:697]: spooked about order of operations stuff here, logically it should go `(mole_count * ratio)/6` but parenthesis are like a blanket to me, do you think we should add them here?

**Review by san7890 — CHANGES_REQUESTED** [review]
buncha nits, but some important things

File: code/modules/surgery/organs/lungs.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:165]: haha yeah I totally was lol lmao

File: code/modules/surgery/organs/lungs.dm (hunk 4 above)
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:192]: falsey args would be invalid too (isnull() is a spook and I don't like it but I'll do it if you really want me to)

File: code/modules/surgery/organs/lungs.dm (hunk 5 above)
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:202]: nope I forgot thanks

File: code/modules/surgery/organs/lungs.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:493]: lmao

File: code/modules/surgery/organs/lungs.dm (hunk 8 above)
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:697]: * and / are commutative, and this isn't a case where floating point shit could butt bite us. So even if it wasn't L->R we'd be ok.
I can do it if you'd like but I think it's fine.

File: code/modules/surgery/organs/lungs.dm (hunk 4 above)
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:192]: fucking mothbloaweddsjdjawds grumble grumble

File: code/modules/surgery/organs/lungs.dm (hunk 8 above)
Comment (san7890) [reply code/modules/surgery/organs/lungs.dm:694]: if you think it's fine then it's probably alright but i personally just don't like fucking around with this stuff in my code so i thought i should mention this

File: code/modules/surgery/organs/lungs.dm [hunk 9]
```diff
@@ -845,10 +877,15 @@
...
-	breath_gases[/datum/gas/water_vapor][MOLES] -= gas_breathed
+/obj/item/organ/internal/lungs/ethereal/Initialize(mapload)
+	. = ..()
+	add_gas_reaction(/datum/gas/water_vapor, while_present = PROC_REF(consume_water))
+
+/// H2O electrolysis
+/obj/item/organ/internal/lungs/ethereal/proc/consume_water(mob/living/carbon/breather, datum/gas_mixture/breath, h2o_pp, old_h2o_pp)
+	var/gas_breathed = breath.gases[/datum/gas/water_vapor][MOLES]
+	breath.gases[/datum/ga
+	s/water_vapor][MOLES] -= gas_breathed
```
Comment (san7890) [inline code/modules/surgery/organs/lungs.dm:888]:
```suggestion
	breath.gases[/datum/gas/water_vapor][MOLES] -= gas_breathed
```

fix your compile (smiley face)

Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:888]: blastr
